src/pylife/mesh/surface.py

```diff
@@ -95,58 +88,66 @@ def _solid_angle(self, x0, y0, z0, x1, y1, z1, x2, y2, z2, x3, y3, z3):
-        df_nodes = df.copy()
-        df_nodes = df_nodes.merge(df_nodes, on="element_id", how="outer", suffixes=["_n0", "_n1"])
-        df_nodes = df_nodes[df_nodes["node_id_n0"] != df_nodes["node_id_n1"]]
-
-        df_nodes = df_nodes.merge(df, on="element_id", how="outer").copy()
-        df_nodes = df_nodes[df_nodes["node_id_n1"] < df_nodes["node_id"]]
-        df_nodes = df_nodes.rename(columns={"node_id": "node_id_n2", "x": "x_n2", "y": "y_n2", "z": "z_n2"})
-
-        df_nodes = df_nodes.merge(df, on="element_id", how="outer", suffixes=["_n2", "_n3"]).copy()
-        df_nodes = df_nodes[df_nodes["node_id_n2"] < df_nodes["node_id"]]
-        df0 = df_nodes.rename(columns={"node_id": "node_id_n3", "x": "x_n3", "y": "y_n3", "z": "z_n3"})
-
-        df0.loc[:,"E"] \
-            = self._solid_angle(df0.x_n0, df0.y_n0, df0.z_n0, df0.x_n1, df0.y_n1, df0.z_n1, df0.x_n2, df0.y_n2, df0.z_n2, df0.x_n3, df0.y_n3, df0.z_n3)
+        df0 = (
+            df.merge(df, on="element_id", how="outer", suffixes=["_n0", "_n1"])
+            .query("node_id_n0 != node_id_n1")
+            .merge(df, on="element_id", how="outer")
+            .query("node_id_n1 < node_id")
+            .rename(
+                columns={"node_id": "node_id_n2", "x": "x_n2", "y": "y_n2", "z": "z_n2"}
+            )
+            .merge(df, on="element_id", how="outer", suffixes=["_n2", "_n3"])
+            .query("node_id_n2 < node_id")
+            .rename(
+                columns={"node_id": "node_id_n3", "x": "x_n3", "y": "y_n3", "z": "z_n3"}
+            )
+            .pipe(lambda df: df.assign(E=self._solid_angle(df)))
+        )
 
         max_E = df0.groupby(["element_id", "node_id_n0"]).max()["E"]
 
-        df1 = df0.join(max_E, on=["element_id", "node_id_n0"], how="left", rsuffix="_max")
-        df2 = df1[df1["E"] == df1["E_max"]]
-        df2 = df2.groupby(["element_id", "node_id_n0"]).first()
-
-        index_columns = df2.reset_index()
-        index_columns = index_columns[["element_id", "node_id_n0"]]
-        df2.index = pd.MultiIndex.from_frame(index_columns, names=["element_id", "node_id"])
-
-        df3 = df2[["x_n0", "y_n0", "z_n0", "E"]]
+        df1 = (
+            df0.join(max_E, on=["element_id", "node_id_n0"], how="left", rsuffix="_max")
+            .query("E == E_max")
+            .groupby(["element_id", "node_id_n0"])
+            .first()
+            .reset_index()
+            .rename(columns={"node_id_n0": "node_id"})
+            .set_index(["element_id", "node_id"])
+        )
 
-        df3 = df3.join(df3.groupby(["node_id"]).sum().rename(columns={"E":"Esum"})["Esum"])
-        df3.loc[:,"is_at_surface"] = df3["Esum"] < 4*np.pi-1e-5
+        df3 = df1[["x_n0", "y_n0", "z_n0", "E"]].join(
+            df1.groupby(["node_id"]).sum().rename(columns={"E": "Esum"})["Esum"]
+        )
 
-        df3["is_at_surface"].fillna(False, inplace = True)
+        df3.loc[:, "is_at_surface"] = df3["Esum"] < 4*np.pi-1e-5
 
```

Comment:
can this be safely omitted? the `fillna`

Reply:
I think so. Look at the following test.

```
Python 3.11.5 (main, Sep 11 2023, 13:54:46) [GCC 11.2.0]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.21.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import pandas as pd

In [2]: import numpy as np

In [3]: pd.Series([1.0, np.nan, 5.0]) < 3.0
Out[3]: 
0     True
1    False
2    False
dtype: bool

In [4]: 
```

So if `df3["Esum"]` is NaN the `<` operator will return false.